graphwalker-core/src/main/java/org/graphwalker/core/algorithm/ChinesePostmanProblem.java

```diff
@@ -0,0 +1,447 @@
+package org.graphwalker.core.algorithm;
+
+import java.util.*;
+
+import org.graphwalker.core.machine.Context;
+import org.graphwalker.core.model.Element;
+import org.graphwalker.core.model.Model;
+import org.graphwalker.core.model.Path;
+import org.graphwalker.core.model.Edge;
+import org.graphwalker.core.model.Vertex;
+
+import java.util.HashMap;
+import java.util.Map;
+
+import static org.graphwalker.core.model.Edge.RuntimeEdge;
+import static org.graphwalker.core.model.Vertex.RuntimeVertex;
+
+/**
+ * Directed Chinese Postman Problem
+ * Algorithm based on: Edmonds, J., Johnson, E.L. Matching, Euler tours and the Chinese postman, Mathematical
+ * Programming (1973) 5: 88. doi:10.1007/BF01580113
+ * @author Onur Enginer
+ */
+public class ChinesePostmanProblem implements Algorithm {
+
+  private final Context context;
+  private final Map<RuntimeVertex, PolarityCounter> polarities;
+  private final Model model;
+  RuntimeVertex neg[], pos[]; // unbalanced vertices
+
+  public ChinesePostmanProblem(Context context) {
+    this.context = context;
+    int N = context.getModel().getVertices().size();
+    this.polarities = new HashMap<>(N);
+    this.model = new Model(context.getModel());
+    polarize();
+  }
+
+  public enum EulerianType {
+    EULERIAN, SEMI_EULERIAN, NOT_EULERIAN
+  }
+
+  private void polarize() {
+    for (RuntimeEdge edge : context.getModel().getEdges()) {
+      getPolarityCounter(edge.getSourceVertex()).decrease();
+      getPolarityCounter(edge.getTargetVertex()).increase();
+    }
+    for (RuntimeVertex vertex : context.getModel().getVertices()) {
+      if (!polarities.get(vertex).hasPolarity()) {
+        polarities.remove(vertex);
+      }
+    }
+  }
+
+  private PolarityCounter getPolarityCounter(RuntimeVertex vertex) {
+    if (!polarities.containsKey(vertex)) {
+      polarities.put(vertex, new PolarityCounter());
+    }
+    return polarities.get(vertex);
+  }
+
+  public Path<Element> getEulerPath(Element element) {
+    Eulerize();
+    return context.getAlgorithm(Fleury.class).getTrail(element);
+  }
+
+  public void Eulerize() {
+    List<Edge> newEdges = new ArrayList<Edge>();
+    int matrixSize = getMatrixSize();
+    int[][] costMatrix = new int[matrixSize][matrixSize];
+    VertexPair[][] pairMatrix = new VertexPair[matrixSize][matrixSize];
+    findUnbalanced();
```

Comment:
Not sure what happens here in the findUnbalanced() method, it looks like we process things without saving or acting upon the result, I might also be tired =)

Reply:
I can also see that "balance" the graph, if that would allow a path to be traversed several times, would be needed when we talk about traversing a model with the least amount of steps

Reply:
> Not sure what happens here in the findUnbalanced() method, it looks like we process things without saving or acting upon the result, I might also be tired =)

This was meant to be where we first store negative and positive polarized vertices in their respective arrays neg[] and pos[], but I have realized that I should do that under polarize method, so I moved the operations there.